### PR TITLE
Add message for overlap maptool

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

The avoidIntersection method uses GEOS as backend to perform the operations.
However, our avoidIntersection method does not check if the geometries are valid.
Geos does not perform this verification either, so the resulting geometry may
still overlap a polygon that is to be avoided.

before

overlap mkv

after
overlap_after mkv

First, I use `makeValid` to try to fix the geometry (as used for selectedFeature for example).
Sedonc, I add a new parameter to return if there was an error after convert a geometry to a valid one.

If there was an unfixable error, a message is raised. Maybe, we should also indicate that the operation was performed with an invalid geometry and QGIS used a modified geometry @gioman?

### CODE REVIEW

File: src/core/geometry/qgsgeometryeditutils.cpp [hunk 1]
```diff
@@ -265,7 +267,11 @@ std::unique_ptr<QgsAbstractGeometry> QgsGeometryEditUtils::avoidIntersections( c
       if ( !f.hasGeometry() )
         continue;
 
-      nearGeometries << f.geometry();
+      QgsGeometry geomValid = f.geometry().makeValid();
```
Comment (roya0045) [inline src/core/geometry/qgsgeometryeditutils.cpp:270]: I guess m values are not needed or these geometries are not set anywhere.

**Comment by gioman** [comment]
> Maybe, we should also indicate that the operation was performed with an invalid geometry and QGIS used a modified geometry @gioman?

yes without doubt, at least in my mind.

**Comment by nyalldawson** [comment]
Please rebase against master before pushing more commits to this branch in order to get the newer workflows enabled

**Comment by nyalldawson** [comment]
> yes without doubt, at least in my mind.

I agree -- in fact, I'd lean towards removing the automatic repairing and instead just alerting to user that they need to fix the geometry. makeValid isn't a magic wand and it has a bunch of assumptions and limitations which make it inappropriate to use silently in the background.

**Comment by lbartoletti** [comment]
This is what it may look like when using makeValid automatically and inform the user that he should first fix the geometries.

overlap_after mkv

Let me know if I need to remove the makeValid part. I wasn't going to fix the geometry initially, but I think it may help users in addition to the message.

We should definitely mention this in the documentation.

**Comment by nyalldawson** [comment]
> Let me know if I need to remove the makeValid part. I wasn't going to fix the geometry initially, but I think it may help users in addition to the message.

I think it's safer to remove it -- otherwise we're causing data loss/corruption for users. I'd rather they have the responsibility to fix things themselves in an appropriate way.

File: src/app/gps/qgsgpsinformationwidget.cpp [hunk 2]
```diff
@@ -1322,7 +1322,7 @@ void QgsGpsInformationWidget::mBtnCloseFeature_clicked()
         }
         else if ( avoidIntersectionsReturn == 3 )
         {
-          QgisApp::instance()->messageBar()->pushCritical( tr( "Add Feature" ), tr( "An error was reported during intersection removal." ) );
+          QgisApp::instance()->messageBar()->pushCritical( tr( "Add Feature" ), tr( "The feature has been added, but at least one geometry intersected is invalid. You should fix geometries." ) );
```
Comment (nyalldawson) [inline src/app/gps/qgsgpsinformationwidget.cpp:1325]:
```suggestion
          QgisApp::instance()->messageBar()->pushCritical( tr( "Add Feature" ), tr( "The feature has been added, but at least one geometry intersected is invalid. These geometries must be manually repaired." ) );
```

File: src/app/qgsmaptoolreshape.cpp [hunk 3]
```diff
@@ -192,6 +194,10 @@ void QgsMapToolReshape::reshape( QgsVectorLayer *vlayer )
             vlayer->destroyEditCommand();
             return;
           }
+          if ( res == 3 )
+          {
+            emit messageEmitted( tr( "At least one geometry intersected is invalid. You should fix geometries." ), Qgis::Warning );
```
Comment (nyalldawson) [inline src/app/qgsmaptoolreshape.cpp:199]:
```suggestion
            emit messageEmitted( tr( "At least one geometry intersected is invalid. These geometries must be manually repaired." ), Qgis::Warning );
```

File: src/gui/qgsmaptooldigitizefeature.cpp [hunk 4]
```diff
@@ -341,9 +341,9 @@ void QgsMapToolDigitizeFeature::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
           QgsGeometry featGeom = f->geometry();
           int avoidIntersectionsReturn = featGeom.avoidIntersections( avoidIntersectionsLayers );
           f->setGeometry( featGeom );
-          if ( avoidIntersectionsReturn == 1 )
+          if ( avoidIntersectionsReturn == 3 )
           {
-            //not a polygon type. Impossible to get there
+            emit messageEmitted( tr( "The feature has been added, but at least one geometry intersected is invalid. You should fix geometries." ), Qgis::Warning );
```
Comment (nyalldawson) [inline src/gui/qgsmaptooldigitizefeature.cpp:346]:
```suggestion
            emit messageEmitted( tr( "The feature has been added, but at least one geometry intersected is invalid. These geometries must be manually repaired." ), Qgis::Warning );
```

**Review by nyalldawson — APPROVED** [review]

File: src/gui/qgsmaptooldigitizefeature.cpp (hunk 4 above)
Comment (roya0045) [reply src/gui/qgsmaptooldigitizefeature.cpp]: Is the 'manually' mandatory? I think simply stating that they should be repair is enough.

Comment (nyalldawson) [reply src/gui/qgsmaptooldigitizefeature.cpp]: I think so -- I want to start discouraging users from thinking that the auto repair tools are a magic wand and the correct solution to use. I'd prefer our message to be "you should fix them yourselves, in a way which is appropriate for the work you are conducting and using your knowledge of the features themselves". Hence "manually"!

Comment (roya0045) [reply src/gui/qgsmaptooldigitizefeature.cpp]: Ok then, it's just that sometime when I add new geometries on a dataset without issue I get the processing error. And I'm always puzzled and annoyed what adding a few simply polygons without overlap seemingly invalidate things and the current fixers seem to do the work, whitout any geometry change. (Hence why the other PR). And also somethings processing can create invalid geometries. Dealing with such can be annoying and time consuming...

**Comment by lbartoletti** [comment]
Thanks!

**Pull request merged**